btree/btree.c

```diff
@@ -2571,11 +2558,11 @@ static void level_cleanup(struct m0_btree_oimpl *oi,
 	int i;
 	for (i = 0; i <= oi->i_used; ++i) {
 		if (oi->i_level[i].l_node != NULL) {
-			node_put(oi->i_level[i].l_node);
+			node_put(oi->i_level[i].l_node, false);
```

Comment:
level_cleanup is called from multiple places. It seems better if level_cleanup takes an extra parameter lock_acquired and passes it to the node_put routines. Also the callers of level_cleanup can pass the current state of the lock to level_cleanup properly.

Reply:
We can do that, But current implementation ensures that level_cleanup will be called after unlocking . So we even if we passed parameter for lock_acquired, value for it is going to be false. Because level_cleanup will get called at P_CLEANUP phase, and there can be only two possibility i.e. we might be in LOCKALL  mode or non-LOCKALL, 
If we are not in LOCKALL mode, i.e. either we have completed our operation and we are in unlock mode, and even if we have not completed operation, we have to traversed again, but we will be unlock the lock and go for P_CLEANUP phase. i.e. we will come at P_CLEANUP when we are in unlock state.

Akso, if we are want to go for LOCKALL mode, we are traversing again after unlocking and then calling P_CLENUP and then P_SETUP or P_LOCKALL state (refer P_CHECK phase line no. :3128)

Moreover, 
level_cleanup(oi, bop->bo_tx);
return P_SETUP;
above steps can be return as m0_sm_op_sub(&bop->bo_op, P_CLEANUP, P_SETUP); as it provides more readability and suggested by NIkhil as well as he is also using the m0_sm_op_sub instead of calling level_cleanup() explicitly.

Reply:
I see that you are unlocking just before calling level_cleanup at some places if this unlocking is done just to make sure that level_cleanup is called in unlocked mode then that is not right as in future if we have a code which mistakenly does not unlock before calling level_cleanup then it will fail.

Reply:
The unlocking is not done just to make sure that level_cleanup is called in unlocked mode, it is done because there is no need to lock anymore.

Also, I have discussed this with Nikhil and Yash, they are fine with either way as   get and iter code,not calling level_cleanup in lock mode and same for put and delete

Reply:
Even if that is the case there needs to be a comment in level_cleanup to indicate that this function assumes the thread is unlocked when level_cleanup runs.

Reply:
okay I will add that

Reply:
comment added